diff --git a/lsplant/src/main/jni/art/runtime/art_method.cxx b/lsplant/src/main/jni/art/runtime/art_method.cxx
index 8b7efc9..db8b6e6 100644
--- a/lsplant/src/main/jni/art/runtime/art_method.cxx
+++ b/lsplant/src/main/jni/art/runtime/art_method.cxx
@@ -13,6 +13,9 @@ import hook_helper;
 
 export namespace lsplant::art {
 class ArtMethod {
+    inline static auto SetNotIntrinsic_ =
+        "_ZN3art9ArtMethod15SetNotIntrinsicEv"_sym.as<void (ArtMethod::*)()>;
+
     inline static auto PrettyMethod_ =
             "_ZN3art9ArtMethod12PrettyMethodEPS0_b"_sym.as<std::string(ArtMethod::*)(bool)>;
 
@@ -103,6 +106,16 @@ public:
         SetAccessFlags(access_flags);
     }
 
+    void SetNonIntrinsic() {
+        if (SetNotIntrinsic_) [[likely]] {
+            SetNotIntrinsic_(this);
+        } else if (kAccIntrinsic) [[likely]] {
+            auto access_flags = GetAccessFlags();
+            access_flags &= ~kAccIntrinsic;
+            SetAccessFlags(access_flags);
+        }
+    }
+
     bool IsPrivate() { return GetAccessFlags() & kAccPrivate; }
     bool IsProtected() { return GetAccessFlags() & kAccProtected; }
     bool IsPublic() { return GetAccessFlags() & kAccPublic; }
@@ -294,6 +307,11 @@ public:
             kAccPreCompiled = 0x00800000;
         }
         if (sdk_int < __ANDROID_API_Q__) kAccFastInterpreterToInterpreterInvoke = 0;
+        if (sdk_int < __ANDROID_API_O__) kAccIntrinsic = 0;
+
+        if (sdk_int >= __ANDROID_API_P__ && !handler(SetNotIntrinsic_)) {
+            LOGW("Failed to find SetNotIntrinsic, use hard-coded kAccIntrinsic instead");
+        }
 
         if (!handler(GetMethodShortyL_, true, GetMethodShorty_)) {
             LOGE("Failed to find GetMethodShorty");
@@ -367,6 +385,7 @@ private:
     inline static uint32_t kAccPreCompiled = 0x00200000;
     inline static uint32_t kAccCompileDontBother = 0x02000000;
     inline static uint32_t kAccDefaultConflict = 0x01000000;
+    inline static uint32_t kAccIntrinsic = 0x80000000;
 };
 
 }  // namespace lsplant::art
diff --git a/lsplant/src/main/jni/art/runtime/class_linker.cxx b/lsplant/src/main/jni/art/runtime/class_linker.cxx
index 361c630..53371fe 100644
--- a/lsplant/src/main/jni/art/runtime/class_linker.cxx
+++ b/lsplant/src/main/jni/art/runtime/class_linker.cxx
@@ -50,43 +50,45 @@ private:
     inline static auto RegisterNativeThread_ =
         "_ZN3art6mirror9ArtMethod14RegisterNativeEPNS_6ThreadEPKvb"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method, Thread *thread, const void *native_method, bool is_fast) static -> void {
-            return backup(thiz, MayGetBackup(method), thread, native_method, is_fast);
+        (ArtMethod *method, Thread *thread, const void *native_method,
+         bool is_fast) static -> void {
+            return backup(MayGetBackup(method), thread, native_method, is_fast);
         };
 
     inline static auto UnregisterNativeThread_ =
         "_ZN3art6mirror9ArtMethod16UnregisterNativeEPNS_6ThreadE"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method, Thread *thread) static -> void {
-            return backup(thiz, MayGetBackup(method), thread);
+        (ArtMethod *method, Thread *thread) static -> void {
+            return backup(MayGetBackup(method), thread);
         };
 
     inline static auto RegisterNativeFast_ =
         "_ZN3art9ArtMethod14RegisterNativeEPKvb"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method, const void *native_method, bool is_fast) static -> void {
-            return backup(thiz, MayGetBackup(method), native_method, is_fast);
+        (ArtMethod *method, const void *native_method, bool is_fast) static -> void {
+            return backup(MayGetBackup(method), native_method, is_fast);
         };
 
-    inline static auto UnregisterNativeFast_ =
-        "_ZN3art9ArtMethod16UnregisterNativeEv"_sym.hook->*[]
+    inline static auto RegisterNativeFastWithReturn_ =
+        "_ZN3art9ArtMethod14RegisterNativeEPKvb"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method) static -> void{
-            return backup(thiz, MayGetBackup(method));
+        (ArtMethod *method, const void *native_method,
+         bool is_fast) static -> const void * {
+            return backup(MayGetBackup(method), native_method, is_fast);
         };
 
     inline static auto RegisterNative_ =
         "_ZN3art9ArtMethod14RegisterNativeEPKv"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method, const void *native_method) static -> const void * {
-            return backup(thiz, MayGetBackup(method), native_method);
+        (ArtMethod *method, const void *native_method) static -> const void * {
+            return backup(MayGetBackup(method), native_method);
         };
 
     inline static auto UnregisterNative_ =
         "_ZN3art9ArtMethod16UnregisterNativeEv"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, ArtMethod *method) static -> const void * {
-            return backup(thiz, MayGetBackup(method));
+        (ArtMethod *method) static -> void {
+            return backup(MayGetBackup(method));
         };
 
     inline static auto RegisterNativeClassLinker_ =
@@ -99,7 +101,7 @@ private:
     inline static auto UnregisterNativeClassLinker_ =
         "_ZN3art11ClassLinker16UnregisterNativeEPNS_6ThreadEPNS_9ArtMethodE"_sym.hook->*[]
         <MemBackup auto backup>
-        (ClassLinker *thiz, Thread *self, ArtMethod *method) static -> const void * {
+        (ClassLinker *thiz, Thread *self, ArtMethod *method) static -> void {
             return backup(thiz, self, MayGetBackup(method));
         };
 
@@ -178,10 +180,13 @@ public:
 				FixupStaticTrampolines_, FixupStaticTrampolinesRaw_);
         if (!method_FixupStaticTrampolines_hooked) LOGD("method FixupStaticTrampolines is inlined");
 
-        if (!handler(RegisterNativeClassLinker_, RegisterNative_, RegisterNativeFast_,
-                          RegisterNativeThread_) ||
-            !handler(UnregisterNativeClassLinker_, UnregisterNative_, UnregisterNativeFast_,
-                          UnregisterNativeThread_)) {
+        auto register_native_hooked = sdk_int >= __ANDROID_API_O__ && sdk_int < __ANDROID_API_P__
+                                          ? handler(RegisterNativeFastWithReturn_)
+                                          : handler(RegisterNativeClassLinker_, RegisterNative_,
+                                                    RegisterNativeFast_, RegisterNativeThread_);
+        if (!register_native_hooked ||
+            !handler(UnregisterNativeClassLinker_, UnregisterNative_,
+                     UnregisterNativeThread_)) {
             return false;
         }
 
@@ -196,7 +201,9 @@ public:
         }
 
         if (!handler(SetEntryPointsToInterpreter_)) [[likely]] {
-            if (handler(GetOptimizedCodeFor_, true)) [[likely]] {
+            if (handler(art_quick_to_interpreter_bridge_)) [[likely]] {
+                // Direct symbol resolved from libart.so
+            } else if (handler(GetOptimizedCodeFor_, true)) [[likely]] {
                 auto obj = JNI_FindClass(env, "java/lang/Object");
                 if (!obj) {
                     return false;
@@ -211,7 +218,7 @@ public:
                 // just in case
                 dummy->SetNonNative();
                 art_quick_to_interpreter_bridge_ = GetOptimizedCodeFor_(dummy.get());
-            } else if (!handler(art_quick_to_interpreter_bridge_)) [[unlikely]] {
+            } else {
                 return false;
             }
         }
diff --git a/lsplant/src/main/jni/art/runtime/jit/jit.cxx b/lsplant/src/main/jni/art/runtime/jit/jit.cxx
index 8f84696..27c98f8 100644
--- a/lsplant/src/main/jni/art/runtime/jit/jit.cxx
+++ b/lsplant/src/main/jni/art/runtime/jit/jit.cxx
@@ -11,12 +11,16 @@ import hook_helper;
 
 namespace lsplant::art::jit {
 enum class CompilationKind {
-    kOsr [[maybe_unused]],
-    kBaseline [[maybe_unused]],
-    kOptimized,
+    kUnknown = -1,
 };
 
 export class Jit {
+    inline static auto kOptimized{CompilationKind::kUnknown};
+
+    inline static auto EnqueueBaselineCompilation_ =
+        "_ZN3art3jit3Jit26EnqueueBaselineCompilationEPNS_9ArtMethodEPNS_6ThreadE"_sym
+            .as<void (Jit::*)(ArtMethod *, Thread *)>;
+
     inline static auto EnqueueOptimizedCompilation_ =
         "_ZN3art3jit3Jit27EnqueueOptimizedCompilationEPNS_9ArtMethodEPNS_6ThreadE"_sym.hook->*[]
         <MemBackup auto backup>
@@ -32,7 +36,7 @@ export class Jit {
         "_ZN3art3jit3Jit14AddCompileTaskEPNS_6ThreadEPNS_9ArtMethodENS_15CompilationKindEb"_sym.hook->*[]
         <MemBackup auto backup>
         (Jit *thiz, Thread *self, ArtMethod *method, CompilationKind compilation_kind, bool precompile) static -> void {
-            if (compilation_kind == CompilationKind::kOptimized && !precompile) {
+            if (compilation_kind == kOptimized && !precompile) {
                 if (auto b = IsHooked(method); b) [[unlikely]] {
                     LOGD("Propagate compile task: %p -> %p", method, b);
                     method = b;
@@ -45,6 +49,12 @@ public:
     static bool Init(const HookHandler &handler) {
         auto sdk_int = GetAndroidApiLevel();
 
+        if (sdk_int >= 35 || handler(EnqueueBaselineCompilation_)) [[likely]] {
+            kOptimized = static_cast<CompilationKind>(3);
+        } else {
+            kOptimized = static_cast<CompilationKind>(2);
+        }
+
         if (sdk_int <= __ANDROID_API_U__) [[likely]] {
             handler(EnqueueOptimizedCompilation_);
             handler(AddCompileTask_);
diff --git a/lsplant/src/main/jni/art/runtime/jni/jni_id_manager.cxx b/lsplant/src/main/jni/art/runtime/jni/jni_id_manager.cxx
index 60f4984..0556a51 100644
--- a/lsplant/src/main/jni/art/runtime/jni/jni_id_manager.cxx
+++ b/lsplant/src/main/jni/art/runtime/jni/jni_id_manager.cxx
@@ -6,6 +6,7 @@ export module lsplant:jni_id_manager;
 
 import :art_method;
 import :common;
+import :clazz;
 import :handle;
 import hook_helper;
 
@@ -24,11 +25,24 @@ private:
         return backup(thiz, method);
     };
 
+    inline static auto EncodeGenericIdWithClass_ =
+        ("_ZN3art3jni12JniIdManager15EncodeGenericIdINS_9ArtMethodEEEjNS_6HandleINS_6mirror5ClassEEENS_16ReflectiveHandleIT_EE"_sym |
+         "_ZN3art3jni12JniIdManager15EncodeGenericIdINS_9ArtMethodEEEmNS_6HandleINS_6mirror5ClassEEENS_16ReflectiveHandleIT_EE"_sym)
+            .hook
+            ->*[]<MemBackup auto backup>(JniIdManager *thiz, Handle<mirror::Class> klass,
+                                         ReflectiveHandle<ArtMethod> method) static -> uintptr_t {
+        if (auto *target = IsBackup(method.Get()); target) {
+            LOGD("get generic id for %s", method.Get()->PrettyMethod().c_str());
+            method.Set(target);
+        }
+        return backup(thiz, klass, method);
+    };
+
 public:
     static bool Init(JNIEnv *env, const HookHandler &handler) {
         int sdk_int = GetAndroidApiLevel();
         if (sdk_int >= __ANDROID_API_R__) {
-            if (IsJavaDebuggable(env) && !handler(EncodeGenericId_)) {
+            if (IsJavaDebuggable(env) && !handler(EncodeGenericIdWithClass_, EncodeGenericId_)) {
                 LOGW("Failed to hook EncodeGenericId, attaching debugger may crash the process");
             }
         }
diff --git a/lsplant/src/main/jni/art/runtime/runtime.cxx b/lsplant/src/main/jni/art/runtime/runtime.cxx
index f492fa0..01eb870 100644
--- a/lsplant/src/main/jni/art/runtime/runtime.cxx
+++ b/lsplant/src/main/jni/art/runtime/runtime.cxx
@@ -14,19 +14,10 @@ namespace lsplant::art {
 
 export class Runtime {
 public:
-    enum class RuntimeDebugState {
-        // This doesn't support any debug features / method tracing. This is the expected state
-        // usually.
-        kNonJavaDebuggable,
-        // This supports method tracing and a restricted set of debug features (for ex: redefinition
-        // isn't supported). We transition to this state when method tracing has started or when the
-        // debugger was attached and transition back to NonDebuggable once the tracing has stopped /
-        // the debugger agent has detached..
-        kJavaDebuggable,
-        // The runtime was started as a debuggable runtime. This allows us to support the extended
-        // set
-        // of debug features (for ex: redefinition). We never transition out of this state.
-        kJavaDebuggableAtInit
+    enum class RuntimeDebugState : uint32_t {
+        kNonJavaDebuggable = 0,
+        kJavaDebuggable = 1,
+        kJavaDebuggableAtInit = 2
     };
 
 private:
@@ -91,6 +82,9 @@ public:
 
 export struct JavaDebuggableGuard {
     JavaDebuggableGuard() {
+        if (GetAndroidApiLevel() >= 34) {
+            return;
+        }
         while (true) {
             size_t expected = 0;
             if (count.compare_exchange_strong(expected, 1, std::memory_order_acq_rel,
@@ -113,6 +107,9 @@ export struct JavaDebuggableGuard {
     }
 
     ~JavaDebuggableGuard() {
+        if (GetAndroidApiLevel() >= 34) {
+            return;
+        }
         while (true) {
             size_t expected = 2;
             if (count.compare_exchange_strong(expected, 1, std::memory_order_acq_rel,
diff --git a/lsplant/src/main/jni/lsplant.cc b/lsplant/src/main/jni/lsplant.cc
index 9452d4e..9425bec 100644
--- a/lsplant/src/main/jni/lsplant.cc
+++ b/lsplant/src/main/jni/lsplant.cc
@@ -539,6 +539,8 @@ bool DoHook(ArtMethod *target, ArtMethod *hook, ArtMethod *backup) {
     } else {
         LOGV("Generated trampoline %p", entrypoint);
 
+        target->SetNonIntrinsic();
+
         target->BackupTo(backup);
 
         target->SetNonCompilable();
@@ -830,7 +832,6 @@ using ::lsplant::IsHooked;
 }
 
 [[maybe_unused]] bool MakeDexFileTrusted(JNIEnv *env, jobject cookie) {
-    JavaDebuggableGuard guard;
     if (!cookie) return false;
     return DexFile::SetTrusted(env, cookie);
 }
